### ISSUE TITLE
Standard ObjectManager_adaptor API

### ISSUE THREAD

**Issue by riuriuriu** [Kistler-Group/sdbus-cpp#139]
Hi Stano, I hope you are doing well.

It looks like the `class ObjectManager_adaptor` from StandardInterfaces.h can't really be used to emit InterfacesAdded signals as described in the [org.freedesktop.DBus.ObjectManager](https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager) interface.

Consider e.g. org.bluez:
`org.freedesktop.DBus.ObjectManager` is implemented on the object root `/`. New objects can dynamically pop up below, like: `/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0`. If the previous object is added, we see this signal with gdbus monitor --system --dest org.bluez:

```bash
/: org.freedesktop.DBus.ObjectManager.InterfacesAdded (objectpath '/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0', {..many interfaces here..})
```

To recreate this case, I could try to instantiate a "standalone" object manager on `"/"` and send the signal:

```cpp
auto objectManager = std::make_shared<sdbus::AdaptorInterfaces<sdbus::ObjectManager_adaptor>>(getSdBusConnection(), "/");
m_objectManager->emitInterfacesAddedSignal();
```
The problem here is that `emitInterfacesAddedSignal()` doesn't let me provide the actual object `"/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0"` which should be part of the `org.freedesktop.DBus.ObjectManager.InterfacesAdded` signal. It always sends the own instances object_path`"/"` . There is an overload where you can pass a list of interfaces, but no object_path.

On the other hand, the tutorial says that the `ObjectManager_adaptor` should be added as additional template parameter to AdaptorInterfaces. We would have code like this:

```cpp
class MyAdaptor : public sdbus::AdaptorInterfaces<myinterface_adaptor, sdbus::ObjectManager_adaptor> {
    ///...
};

std::make_shared<MyAdaptor>(getSdBusConnection(), "/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0");
```
But this would also be completely wrong. It would add the ObjectManager interface to the actual managed object. Definitely not what we want.

So is there another way how to use `class ObjectManager_adaptor` which just isn't obvious? Or is it's API broken?

Many Thanks

**Comment by dleeds-cpi** [comment]
`ObjectManager_adaptor` appears to work as expected. Only emit a signal for the actual object that was created. However, I have run into an issue with the `ObjectManager_proxy`, which would generally want to subscribe to InterfacesAdded signals using a `path_namespace` rule.

**Comment by riuriuriu** [comment]
> This is exactly what you want. All objects that may emit ObjectManager interfaces must implement the ObjectManager interface. ObjectManager_adaptor appears to work as expected.

I have to object. Please see [org.freedesktop.DBus.ObjectManager](https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager):
>The root of each sub-tree implements this interface so other applications can get all objects, interfaces and properties in a single method call.

The object manager should be implemented on the root of a sub-tree. The object which sends the signal `org.freedesktop.DBus.ObjectManager.InterfacesAdded` is **not** the same as it's parameter `object_path`. 

Repeating the real-live example from bluez:
```
/: org.freedesktop.DBus.ObjectManager.InterfacesAdded (objectpath '/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0', {..many interfaces here..})
```

The signal sender is on path `/` and the managed object's path is `/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0`. It looks like this can't be done with the current API of the standard `ObjectManager_adaptor`.

**Comment by sangelovic** [comment]
@riuriuriu Hey Urs, is this issue still relevant? If so, are you looking into that, or shall I dive into it?

**Comment by dleeds-cpi** [comment]
@riuriuriu & @sangelovic I reconsidered my response, and I see @riuriuriu's point. The ObjectManager interface should be implemented by a root object, and `Object::emitInterfacesAddedSignal()` and `Object::emitInterfacesRemovedSignal()` should take the object path of a child object as a parameter. Right now the Object and Proxy are not symmetrical.

**Comment by riuriuriu** [comment]
> @riuriuriu Hey Urs, is this issue still relevant? If so, are you looking into that, or shall I dive into it?

Yes it's still an issue. As it will mean a breaking API change I wanted to prepare an RFC/Draft but was still held up by other projects. I can do it in the last week of August. If that's too late, feel free to take over. 

The biggest pain point is `class ObjectManager_adaptor` in `StandardInterfaces.h`. It wrongfully assumes that the ObjectManager (the object which implements `org.freedesktop.DBus.ObjectManager`) and the actual managed objects are the same. But they are not. In the bluez example above, the ObjectManager is on `/` but managed objects are on paths like `/org/bluez/hci0/dev_5B_3D_86_7A_D1_A0`.

These are the possibilities I considered:

1. As @dleeds-cpi also suggested, Object::emitInterfacesAddedSignal() and Object::emitInterfacesRemovedSignal() and also the adaptor could take an object path argument for the managed object. Nevertheless I find this not so convenient as it makes things more complicated than the underlying sdbus library. In sdbus you can i.e. just `sd_bus_emit_object_added(sd_bus *bus, const char *path);`. So you also specify the path of the managed object. But the closest registered ObjectManager in the path above is automatically found and the signal is sent from there.
2. Split the `class ObjectManager_adaptor` into two classes: `class ObjectManager_adaptor` and `class ManagedObject_adaptor`. The `class ObjectManager_adaptor` would just register the object manager at the underlying sdbus library. The `class ManagedObject_adaptor` would then provide methods like `emitInterfacesAddedSignal()` and use it's own path. No additional path argument required as in (1).
3. Same as (2) but emiting the `InterfacesAdded` and `InterfacesRemoved` signals automatically on construction and destruction of the `ManagedObject`. On the convenience layer, the set of interfaces are known at compile time as they are template arguments to `class AdaptorInterfaces`.

My favorite is to go with option (3). If (3) seems to tricky to implement, then with (2).

**Comment by dleeds-cpi** [comment]
@riuriuriu (2) and (3) are clever and would satisfy the majority use case. I do have a couple of concerns about them, which we should consider before pursing them.
1. While rarely used, it is possible to signal adding and removing specific interfaces after an object has been created. This is why there are overloads for `emitInterfacesAdded()` and `emitInterfacesRemoved()` that take specific interfaces. This could be supported by option (2), but not so easily by option (3).
2. Adaptor and Proxy objects are conventionally created symmetrically. If we created a `ManagedObject_adaptor`, there wouldn't be an analogous `ManagedObject_proxy`, since everything would be done by the root object's `ObjectManager_proxy`. This could be confusing.
